cmd/snap-device-helper/snap-device-helper.c

```diff
@@ -143,24 +143,38 @@ static char *udev_to_security_tag(const char *udev_tag) {
     return tag;
 }
 
-/* sysroot can be mocked in tests */
-static const char *sysroot = "/";
-
 int snap_device_helper_run(const struct sdh_invocation *inv) {
     const char *action = inv->action;
     const char *udev_tagname = inv->tagname;
-    const char *devpath = inv->devpath;
-    const char *majmin = inv->majmin;
+    const char *major = inv->major;
+    const char *minor = inv->minor;
+    const char *subsystem = inv->subsystem;
 
     bool allow = false;
 
-    if (strlen(majmin) < 3) {
-        die("no or malformed major/minor \"%s\"", majmin);
+    if ((major == NULL) && (minor == NULL)) {
+        /* no device node */
+        return 0;
+    }
+    if ((major == NULL) || (minor == NULL)) {
+        die("incomplete major/minor");
     }
-    if (strlen(devpath) <= strlen("/devices/")) {
-        die("no or malformed devpath \"%s\"", devpath);
+    if (subsystem != NULL) {
+        /* ignore kobjects that are not devices */
+        if (strcmp(subsystem, "subsystem") == 0) {
+            return 0;
+        }
+        if (strcmp(subsystem, "module") == 0) {
+            return 0;
+        }
+        if (strcmp(subsystem, "drivers") == 0) {
+            return 0;
+        }
     }
-    if (sc_streq(action, "bind") || sc_streq(action, "add") || sc_streq(action, "change")) {
+
+    if (action == NULL) {
+        die("ERROR: no action given");
+    } else if (sc_streq(action, "bind") || sc_streq(action, "add") || sc_streq(action, "change")) {
```

Comment:
Since `die()` will exit if action is NULL, would it be better to remove the `else` on this `if` clause?